support YAML format comment

k-kinzal opened the pull request:
Support YAML format comment.

In the helm chart, detailed structures may be commented.

```yaml
securityContext: {}
  # capabilities:
  #   drop:
  #   - ALL
  # readOnlyRootFilesystem: true
  # runAsNonRoot: true
  # runAsUser: 1000
```

So, if you have an empty array or an empty map, you can get the detailed structure by commenting out the YAML when generating the schema.

## disable flag

```json
{
    "securityContext": {
        "type": "object"
}
```

## enable flag

```json
{
    "securityContext": {
        "type": "object",
        "properties": {
            "capabilities": {
                "type": "object",
                "properties": {
                    "drop": {
                        "type": "array",
                        "items": {
                            "type": "string"
                        }
                    }
                }
            },
            "readOnlyRootFilesystem": {
                "type": "boolean"
            },
            "runAsNonRoot": {
                "type": "boolean"
            },
            "runAsUser": {
                "type": "integer"
            }
        }
    }
```

k-kinzal committed:
support YAML format comment

karuppiah7890 commented:
@k-kinzal Wow! Thanks for this contribution! I have not worked with YAML comments, I have only seen people do it in other tools like helm-docs etc. This is super awesome! :)

I was just pulling the code in my local and trying it out manually to test it out. It worked with the testdata present in the repo, which is what you have also mentioned in the PR comments.

I was reading through the code but I had some difficulty in understanding it. So, I have a few questions since I have not done anything related to YAML comments before

- I'm assuming the `commentOutYAML` is uncommenting the comments that have some different valid example YAML which is different from what's present in the YAML key, value content. Am I right? Could we name the function as `uncommentYAML` maybe? Since "commenting out" sounds like you are adding comments or converting code to comments, what do you think?
- What is the logic that is used to uncomment the YAML? I see there's recursion being used and based on the kind of YAML node, some processing is done. But I'm still not able to exactly understand what's going on 😅
- Looks like we only consider Foot Comments, that is comments after the key, value content? Correct?
- By manual testing, I found out that the comments which do not contain YAML are not considered and just ignore. How is this taken care of? The data I used for testing is this

```yaml
securityContext: {}
  # blah blah bloo bloo
  # capabilities:
  #   drop:
  #   - ALL
  # readOnlyRootFilesystem: true
  # runAsNonRoot: true
  # runAsUser: 1000
```

So, in this case, it doesn't have any JSON Schema for `securityContext` as the foot note comment has simple english comments and then only some YAML content

- I also see some duplication of code a bit, just by blind review of code. Could we make it a bit better?
- Could we write some unit test or just integration test or any test to check this feature? I guess I'm okay with even end to end test to automate what I just tested - manually running and testing.

k-kinzal commented:
@karuppiah7890 
Thanks for the reply!!!

> I'm assuming the commentOutYAML is uncommenting the comments that have some different valid example YAML which is different from what's present in the YAML key, value content. Am I right? Could we name the function as uncommentYAML maybe? Since "commenting out" sounds like you are adding comments or converting code to comments, what do you think?

Yes, you're right, I think `uncommentYAML` is easy to understand.
Add the commit I changed to `commentYAML`.

> What is the logic that is used to uncomment the YAML? I see there's recursion being used and based on the kind of YAML node, some processing is done. But I'm still not able to exactly understand what's going on 😅

There are five types of nodes in yaml.v3.

https://github.com/go-yaml/yaml/blob/v3/yaml.go#L325-L331
- DocumentNode - YAML root node
- SequenceNode 
- MappingNode 
- ScalarNode
- AliasNode

Since there are three types of nodes with children, `DocumentNode`, `SequenceNode`, and `MappingNode`, we will explore them recursively.

Replace child elements with nodes that match the condition during recursion as YAML comments.
- `SequenceNode` or `MappingNode`
- Child is empty
- `FootComment` exists
- `FootComment` can be unmarshal as YAML
- Node types match between parent and child


I was writing a comment and found that the code was wrong.
I will fix it to a simpler and easier to understand code.


> Looks like we only consider Foot Comments, that is comments after the key, value content? Correct?

Yes, that's correct.
The comment in map looks like this.

```yaml
# head comment foo
foo: # line comment foo
  # head comment bar
  bar: {} # line comment bar
  # foot comment bar
# foot comment foo
```

> By manual testing, I found out that the comments which do not contain YAML are not considered and just ignore. How is this taken care of? The data I used for testing is this

https://github.com/karuppiah7890/helm-schema-gen/pull/10/files#diff-ab967ab1a2f3a1b769106eeb7bfe892ef0e81d1d27811fa15be08e6749feee1fR32-R34

In this case, the `FootComment` will be commented.

```
# blah blah bloo bloo
# capabilities:
#   drop:
#   - ALL
# readOnlyRootFilesystem: true
# runAsNonRoot: true
# runAsUser: 1000
```

Replace `#` in this comment and give it an indent to make it an uncommented string.

```
 blah blah bloo bloo
 capabilities:
   drop:
   - ALL
 readOnlyRootFilesystem: true
 runAsNonRoot: true
 runAsUser: 1000
```

This string is parsed as YAML and fails, so it is detected as not being a comment in YAML format.


> - I also see some duplication of code a bit, just by blind review of code. Could we make it a bit better?
> - Could we write some unit test or just integration test or any test to check this feature? I guess I'm okay with even end to end test to automate what I just tested - manually running and testing.

OK! I'll fix it.

k-kinzal committed:
rename commentOutYAML -> uncommentYAML

k-kinzal committed:
refactor uncommentYAML && add test

k-kinzal committed:
fixes a case where HeadComment is used

```yaml
ingress:
  enabled: false
  annotations: {}
    # kubernetes.io/ingress.class: nginx
    # kubernetes.io/tls-acme: "true"
  hosts: []
```

In this case, `HeadComment` in `hosts` is used.

k-kinzal committed:
add pattern test

k-kinzal commented on cmd/root.go:
```diff
@@ -2,14 +2,62 @@ package cmd
+				}
+			}
+
+			if i > 0 && len(key.HeadComment) > 0 {
+				prev := node.Content[i-1]
+				if len(prev.Content) == 0 && (prev.Kind == yaml.MappingNode || prev.Kind == yaml.SequenceNode) {
+					comment := strings.ReplaceAll(key.HeadComment, "#", strings.Repeat(" ", key.Column-1))
+					root := yaml.Node{}
+					if err := yaml.Unmarshal([]byte(comment), &root); err == nil {
+						if root.Content[0].Kind == prev.Kind {
+							node.Content[i].HeadComment = ""
+							node.Content[i-1].Content = append(prev.Content, root.Content[0].Content...)
+						}
+					}
+				}
+			}
```
```yaml
ingress:
  enabled: false
  annotations: {}
    # kubernetes.io/ingress.class: nginx
    # kubernetes.io/tls-acme: "true"
  hosts: []
```

In this case, `HeadComment` in `hosts` node is used.

k-kinzal commented on testdata/values.uncommented.yaml:
```diff
@@ -0,0 +1,79 @@
+  minReplicas: 1
+  maxReplicas: 100
+  targetCPUUtilizationPercentage: 80
+  # targetMemoryUtilizationPercentage: 80
```
Uncommenting this case is very difficult in `yaml.v3`.
https://play.golang.org/p/9yWu8YtMCZ1
https://play.golang.org/p/aFO9ukr7r7H

You need to look at the indentation to see which level to add it to.


I think we should support this case.
However, supporting this case makes the code more complex.

k-kinzal commented:
@karuppiah7890 
I fixed it.

However, it needs to be discussed.
https://github.com/karuppiah7890/helm-schema-gen/pull/10#discussion_r575694921
I think it is difficult to uncomment all YAML format comments.
We need to choose whether to support only partial comments or not to support commenting out of this YAML format.

karuppiah7890 commented:
@k-kinzal I didn't get that. 😅

Also, looks like you foresee possible problems with this feature? If so what are they? Could you show them with examples?

k-kinzal commented:
@karuppiah7890 
I have created a code that uncomments an empty array, an empty object in YAML format.

```yaml
foo:
# bar: 1
# baz: 2
```

This case is not a problem.

If you want to uncomment YAML comments in an array or an object, there are a few difficult cases.

```yaml
# pattern 1
foo:
  bar: {}
# baz: {} # this foobar HeadComment
foobar: 1
```

```yaml
# pattern 2
foo:
  bar: {}
  # baz: {} # this bar FootComment
foobar: 1
```

```yaml
# pattern 3
foo:
  bar: {}
    # baz: {} # this bar FootComment
foobar: 1
```

```yaml
# pattern 4
foo:
  bar: {}
#    baz: {} # this foobar HeadComment
foobar: 1
```

These are indented differently.
The problem is as follows

- Where should the `baz` be inserted?
- If the insert destination is different between pattern 2 and pattern 3, it cannot be written in the code because `yaml.v3` cannot get the indentation of the comment.
- If we insert `baz` into `bar` in the last pattern, then we can't reference `bar` in the current `uncommentYAML`.

karuppiah7890 commented:
Hey @k-kinzal . Sorry, I have been pretty caught up with some things and didn't have the bandwidth to look into this. Seems like a really hard problem. I still have to research into this. * in a sleepy state as of now 😅 🙈 * What do you suggest we do? Should we still provide the feature with a feature toggle as you have put - the flag, so that people can still use the feature in some cases and then we can iterate on it later? To solve problems that you have spoken about. I'll also re-read and understand and think about the problems you mentioned

karuppiah7890 commented:
I feel really bad for not merging this PR. It's a super cool magical feature. I feel like the gate keeper who's stopping good features from getting in. I just wanted to have a discussion. Maybe we can merge this, after you do a self review. I still am trying to get my head around it I think.

karuppiah7890 commented:
And then we can release it :)

k-kinzal commented:
@karuppiah7890 
Thanks! I have come up with another approach, may I close this PR?

If we stop changing node in YAML and change it in strings or adopt [goccy/go-yaml](https://github.com/goccy/go-yaml), we may be able to solve the problem.

karuppiah7890 commented:
Okay! 👍
